### PR TITLE
496 load mapper func

### PR DESCRIPTION
**Pull request by jwestw**
## Pull Request submission

This PR focuses on improving the loading and validation of mapper data.

A new function `getmappername(mapper_path_key, split)` has been created. This function uses a regular expression to extract the mapper name from the mapper_path_key. If the split argument is True, it replaces underscores in the mapper name with spaces.

A new function `load_valdiate_mapper(...)` has been created. This function is the main thing in this PR. It loads a specified mapper, validates it using a schema and an optional secondary validation function. The function takes several arguments including a mapper path key, a dictionary of paths, a function to check if a file exists, a function to read a CSV file into a DataFrame, a logger, a function to validate a DataFrame against a schema, an optional function to perform additional validation on the DataFrame, and any additional arguments to pass to the validation function. The function returns the loaded and validated mapper DataFrame.

#### Closes or fixes
RDRP-496
* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/staging/staging_main.py [hunk 1]
```diff
@@ -466,34 +549,37 @@ def run_staging(
...
+        paths,
+        check_file_exists,
+        read_csv,
+        StagingMainLogger,
+        val.validate_data_with_schema,
+        val.validate_ultfoc_df,
+    )
+
+    itl_mapper = load_valdiate_mapper(
+        "ultfoc_mapper_path",
```
Comment (JenCheshire) [inline src/staging/staging_main.py]: This needs changing to the itl mapper (currently in config as ` itl_path`)

**Review by JenCheshire — COMMENTED** [review]
This is great!  Just a path to change for one of the mappers.
Spotted a print statement in line 413 of `staging_main` - it wasn't in the changed code, but do you know if that's intentional? 
I'll have another look in the morning with fresh eyes

File: src/staging/staging_main.py (hunk 1 above)
Comment (jwestw) [reply src/staging/staging_main.py]: I have changed this. Thanks

File: src/staging/staging_main.py [hunk 2]
```diff
@@ -15,6 +16,88 @@
...
+    one_to_many_val_func: Callable,
+    *args,
+):
+    """
+    Loads a specified mapper, validates it using a schema and an optional validation function.
+
+    This function first retrieves the path of the mapper from the provided paths dictionary using the mapper_path_key.
+    It then checks if the file exists at the mapper path. If the file exists, it is read into a DataFrame.
+    The DataFrame is then validated against a schema, which is located at a path constructed from the mapper name.
+    If a validation function is provided, it is called with the DataFrame and any additional arguments.
```
Comment (AnneONS) [inline src/staging/staging_main.py]: Really helpful docstring!

File: src/staging/staging_main.py [hunk 3]
```diff
@@ -15,6 +16,88 @@
...
+        val_with_schema_func (Callable): A function to validate a DataFrame against a schema.
+        validation_func (Callable, optional): An optional function to perform additional validation on the DataFrame.
+        *args: Additional arguments to pass to the validation function.
+
+    Returns:
+        pd.DataFrame: The loaded and validated mapper DataFrame.
+
+    Raises:
+        FileNotFoundError: If no file exists at the mapper path.
+        ValidationError: If the DataFrame fails schema validation or the validation function.
```
Comment (AnneONS) [inline src/staging/staging_main.py]: Yes, we should be adding "Raises" to more of our docstrings.

File: src/staging/staging_main.py [hunk 4]
```diff
@@ -15,6 +16,88 @@
...
+
+    # Validate the DataFrame against the schema
+    val_with_schema_func(mapper_df, schema_path)
+
+    # If a one-to-many validation function is provided, validate the DataFrame
+    if one_to_many_val_func:
+        # Prepend the DataFrame to the arguments
+        args = (mapper_df,) + args
+        # Call the validation function with the DataFrame and the other arguments
+        one_to_many_val_func(*args)  # args include "col_many" and "col_one"
```
Comment (AnneONS) [inline src/staging/staging_main.py]: neat!

File: src/staging/staging_main.py [hunk 5]
```diff
@@ -15,6 +16,88 @@
...
 
+def getmappername(mapper_path_key, split):
+
+    patt = re.compile(r"^(.*?)_path")
+    mapper_name = re.search(patt, mapper_path_key).group(1)
+
+    if split:
+        mapper_name = mapper_name.replace("_", " ")
+
+    return mapper_name
```
Comment (AnneONS) [inline src/staging/staging_main.py]: Nice

**Review by AnneONS — COMMENTED** [review]
Really nice code! Just one question, is there a reason for not validating the cellno mapper in the same way as all the others?

**Comment by jwestw** [comment]
This mapper will change every year. The cell no mapper isn't put through the same function as this 

Also move the new functions outside `staging_main.py` into a new `staging_helpers.py`

**Comment by jwestw** [comment]
> This is great! Just a path to change for one of the mappers. Spotted a print statement in line 413 of `staging_main` - it wasn't in the changed code, but do you know if that's intentional? I'll have another look in the morning with fresh eyes

This should be sorted now in https://github.com/ONSdigital/research-and-development/pull/152/commits/2b3c2036b8baff014be709f0ff2bc5aef9244add

File: src/staging/staging_main.py [hunk 6]
```diff
@@ -245,6 +245,13 @@ def run_staging(
         val.validate_data_with_schema,
         val.validate_ultfoc_df,
     )
+    # # Debug begin
+    # cora_mapper = cora_mapper.astype("str")
+    # StagingMainLogger.info(f"Cora mapper coulms:\n{cora_mapper.dtypes}")
+    # # Debug end
+
+    # cora_mapper = val.validate_cora_df(cora_mapper)
+    # StagingMainLogger.info("Cora status mapper file loaded successfully...")
```
Comment (JenCheshire) [inline src/staging/staging_main.py:254]: Assume this needs to be taken out (all refs to cora mapper will taken out with 574)?

**Review by JenCheshire — COMMENTED** [review]
Looking good

**Review by AnneONS — APPROVED** [review]
Code reviewed and tested, happy to merge.

**Pull request merged**